posts/2024-05-22-multi-dim-arrays.qmd

```diff
@@ -0,0 +1,283 @@
+dimension has size `7`. Now, however, using F-order, the *first* dimension will have stride `1`, the *second* dimension will
+have stride `3`, and the *third* dimension will have stride `3*5 = 15`.
+
+## Dimension semantics
+
+The dimensions of a multi-dimensional array sometimes come with additional semantics depending 
+on what data they store. We discuss interpreting arrays as matrices and images.
+
+### Matrices
+
+Matrices are often represented as a 2D array of numbers. Horizontal groupings of these numbers are called "rows" and vertical
+groupings are called "columns." In mathematics, the entries of a matrix $A$ are denoted $a_{ij}$. 
+
+**Recommendation:** Software should be clearly communicate when arrays represent matrices, and follow the *Matrix indexing
+convention*. Software and documentation should use the terms "row-major" and "column-major" only when referring to matrices.
+
+::: {.callout-note appearance="minimal"}
+# Matrix indexing convention
+
+The first index of a matrix ($i$) refers to rows, the second index ($j$) refers to columns.
+:::
+
+
+::: {.callout-tip collapse=true appearance="minimal"}
+# Row- and column-major
+
+The terms row- and column-major derive for the storage of matrices. Defining these terms first depends on first agreeing which
+index (first or last) refers to rows vs columns for matrices in mathematics.
+
+* **Consequence:** Given matrix indexing conventions, C-order storage is equivalent to "row-major".
+* **Consequence:** Given matrix indexing conventions, F-order storage is equivalent to "column-major".
+
+#### example 
+
+As a result of the *Matrix Indexing Convention* the size of a matrix with `2` rows and `3` columns is `[2, 3]` for both C- and
+F-orderings.  Consider:
+
+```
+          column 0   column 1   column 2
+  row 0  [    0          1          2   ]
+  row 1  [    3          4          5   ]
+```
+
+* The flat C-ordered array will be: `[0, 1, 2, 3, 4, 5]`
+* The flat F-ordered array will be: `[0, 3, 1, 4, 2, 5]`
+
+To reiterate, the multi-dimensional indexes for both C- and F-order are:
+
+```
+          column 0   column 1   column 2
+  row 0  [  (0,0)     (0,1)       (0,2) ]
+  row 1  [  (1,0)     (1,1)       (1,2) ]
+```
+
+because, the row index is *always* the first index.
+
+:::
+
+For matrices, C- and F- order indexing will agree on the ordering of an array's indexes and dimensions, but will store the
+arrays differently when flattened. This is because the matrix indexing convention attaches semantics (row/column) to
+the index position (first/second).
+
+### Images
+
+Two-dimensional images are often stored as arrays where two dimensions vary the horizontal and vertical positions of the
+samples, and as a result these dimensions should be displayed horizontally and vertically, respectively.  Most formats for
+storing "natural" images store data such that the "horizontal axis" / rows have a smaller stride than the "vertical axis" /
```

Comment:
what is a natural image?

Reply:
Tough to define precisely, but roughly  "an image of physical objects that a human might see using the unaided eye"  It's a [pretty common term in computer vision](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C21&q=%22natural+image%22+processing&btnG=&oq=nat)

Reply:
might be useful here to clarify that "horizontal" and "vertical" are relative to the camera sensor, not the subject. Landscape and portrait images of the same subject should probably not be displayed based on the memory layout.